lib/go-parser/index.ts

```diff
@@ -0,0 +1,104 @@
+import * as elf from "elfy";
+import * as path from "path";
+import { Readable } from "stream";
+import { ExtractAction } from "../extractor/types";
+
+const ignoredPaths = [
+  path.normalize("/boot"),
+  path.normalize("/dev"),
+  path.normalize("/etc"),
+  path.normalize("/home"),
+  path.normalize("/media"),
+  path.normalize("/mnt"),
+  path.normalize("/proc"),
+  path.normalize("/root"),
+  path.normalize("/run"),
+  path.normalize("/sbin"),
+  path.normalize("/sys"),
+  path.normalize("/tmp"),
+  path.normalize("/var"),
+];
+
+function filePathMatches(filePath: string): boolean {
+  const dirName = path.dirname(filePath);
+  return (
+    !path.parse(filePath).ext &&
+    !ignoredPaths.some((ignorePath) => dirName.startsWith(ignorePath))
+  );
+}
+
+export const getGoModulesContentAction: ExtractAction = {
+  actionName: "gomodules",
+  filePathMatches,
+  callback: findGoBinaries,
+};
+
+async function findGoBinaries(stream: Readable): Promise<any> {
+  return new Promise((resolve, reject) => {
+    const encoding = "binary";
+    const buildIdMagic = "Go";
+    const elfHeaderMagic = "\x7FELF";
+    const buildInfoMagic = "\xff Go buildinf:";
+
+    const result: Buffer[] = [];
```

Comment:
Wondering if we could just allocate a single buffer and push incoming bytes to it instead of having an array of buffers? This way we allocate the memory only once, otherwise we do it 2 times: once to build the array and once to concat all the buffers into a new buffer

Reply:
I've done a bit of research and because you can't really create a buffer of dynamic size (you have to provide size) and you can't change the size later either it might not be so straight forward. 
So we would either need to create a buffer of a size that should always be enough fo the binary data we need (for this PR it could be small, because we're only processing on the headers but the next PR we're moving to parse other sections which are further in the file) or we would need to create a buffer of a certain size and if we can't fit more into that buffer, create another buffer, not sure! 🤔 Wdyt?

Reply:
Is there anything in the `tar-stream` extraction where it gives us the "headers" that can tell us the expected file size? https://github.com/snyk/snyk-docker-plugin/blob/master/lib/extractor/layer.ts#L22 Something like `headers.fileSize`? Maybe we could pre-allocate a buffer of that size?

I think the risk here is loading some massive go binary in memory twice - one for each chunk of buffer of the file and another for concat()-ing all chunks.

> or we would need to create a buffer of a certain size and if we can't fit more into that buffer, create another buffer, not sure

I think something along the lines could work! Or just keep only 1 buffer and every time you receive a chunk you can concat it to the existing buffer. The issue here would be lots of allocations so may not be a great option...

I would say do what you think works OK as a first iteration, measure it, and improve later!

Reply:
```
    stream.once("data", (chunk) => {
      const first4Bytes = chunk.toString(encoding, 0, 4);

      if (first4Bytes === elfHeaderMagic) {
        // result.push(Buffer.from(chunk));
        buffer = Buffer.concat([buffer, Buffer.from(chunk)]);
        // Listen to next chunks only if it's an ELF executable
        stream.addListener(
          "data",
          (chunk) => (buffer = Buffer.concat([buffer, Buffer.from(chunk)])),
        );
      }
    });
```
Could do something like above, starting with: 
```
    let buffer: Buffer = new Buffer(0);
```
But, like you said, there's a lot of assignments, I'm not sure how to understand which is better option, I think this will use less memory but there's more expensive operations, what do you think?

Reply:
It was a nice an easy change to do this: 
```
Buffer.from(chunk).copy(buffer, bytesWritten, 0);
        bytesWritten += chunk.length;
```
BUT the `streamSize` I am passing can be undefined (see latest commit), so I was trying to investigate another way to find the size but we can't from the stream and I don't think we know how much file we want to parse. But maybe I can check with @dkontorovskyy 🤔 Wdyt, if you have any ideas, shoot!